### PR TITLE
Styles Workflow - Sampler HUD

### PR DESCRIPTION
**Pull request by volfied**
With sampler tool selected, pressing Spacebar will pop up a HUD with sampling possibilities shown.

Current possibilities are:
Sample fill color / stroke color / type style / layer effects / graphics (between smart objects)

### CODE REVIEW

File: src/js/actions/layereffects.js [hunk 1]
```diff
@@ -337,11 +337,13 @@ define(function (require, exports) {
      * Duplicates the layer effects of the source layer on all the target layers
      *
      * @param {Document} document
-     * @param {Immutable.Iterable.<Layer>} targetLayers
+     * @param {Immutable.Iterable.<Layer>?} targetLayers Default is selected in the document
```
Comment (mcilroyc) [inline src/js/actions/layereffects.js]: ? goes at the beginning

File: src/js/actions/layereffects.js [hunk 2]
```diff
@@ -408,6 +410,8 @@ define(function (require, exports) {
             coalesce: false
         };
 
+        this.dispatch(events.style.HIDE_HUD);
```
Comment (mcilroyc) [inline src/js/actions/layereffects.js]: I think the rule now is that dispatches must use `dispatchAsync` *unless* it is already inside a promise chain.  @iwehrman correct me if I'm misremembering our recent conversation.

File: src/js/actions/sampler.js [hunk 3]
```diff
@@ -176,7 +289,64 @@ define(function (require, exports) {
     };
     click.reads = [locks.PS_DOC, locks.JS_UI];
     click.writes = [];
-    click.transfers = [layerFXActions.duplicateLayerEffects, shapeActions.setFillColor, typeActions.setColor];
+    click.transfers = [
+        layerFXActions.duplicateLayerEffects,
+        shapeActions.setFillColor,
+        typeActions.setColor, applyColor
```
Comment (mcilroyc) [inline src/js/actions/sampler.js]: nit: this should be split in to two lines

File: src/js/actions/type.js [hunk 4]
```diff
@@ -524,15 +524,19 @@ define(function (require, exports) {
      * Applies the given text style to target layers
      *
      * @param {Document} document
-     * @param {Immutable.Iterable.<Layer>} targetLayers
+     * @param {Immutable.Iterable.<Layer>?} targetLayers Default is selected layers
```
Comment (mcilroyc) [inline src/js/actions/type.js]: same nit as above, ? at beginning, = at the end

File: src/js/actions/type.js [hunk 5]
```diff
@@ -524,15 +524,19 @@ define(function (require, exports) {
...
      * @return {Promise}
      */
     var applyTextStyle = function (document, targetLayers, style) {
+        targetLayers = targetLayers || document.layers.selected;
+
         var layerIDs = collection.pluck(targetLayers, "id"),
             layerRefs = layerIDs.map(textLayerLib.referenceBy.id).toArray(),
             applyObj = textLayerLib.applyTextStyle(layerRefs, style);
 
+        this.dispatch(events.style.HIDE_HUD);
```
Comment (mcilroyc) [inline src/js/actions/type.js]: same possible gripe as above, wrt dispatchAsync

File: src/js/actions/layereffects.js (hunk 2 above)
Comment (iwehrman) [inline src/js/actions/layereffects.js]: That is indeed the rule.

File: src/js/models/layer.js [hunk 6]
```diff
@@ -529,5 +558,24 @@ define(function (require, exports, module) {
         return this.kind === this.layerKinds.VECTOR;
     };
 
+    /**
+     * Returns the smart object type for smart object layers, null otherwise
+     *
+     * @return {SmartObjectType}
```
Comment (iwehrman) [inline src/js/models/layer.js]: If you're returning null, should be `?SmartObjectType`

File: src/js/models/layer.js [hunk 7]
```diff
@@ -37,6 +37,29 @@ define(function (require, exports, module) {
         Text = require("./text");
 
     /**
+     * Possible smart object types for a layer
```
Comment (iwehrman) [inline src/js/models/layer.js]: Should this be in adapter/lib?

File: src/js/tools/sampler.js [hunk 8]
```diff
@@ -70,13 +85,33 @@ define(function (require, exports, module) {
...
+            flux.actions.sampler.click(currentDocument, event.pageX, event.pageY, event.shiftKey);
+        }
+    };
+
+    /**
+     * Handler for keydown events, installed when the tool is active.
+     *
+     * @param {CustomEvent} event
+     */
+    SamplerTool.prototype.onKeyUp = function (event) {
```
Comment (iwehrman) [inline src/js/tools/sampler.js]: Can this happen on keydown instead? It feels really sluggish due to the fact that it's on key up.

**Comment by iwehrman** [comment]
A few times I accidentally clicked in the hole in the middle of the stroke icon... it's easier to do than you you might think. Maybe we can make the hole slightly opaque so that JS side gets that event?

**Comment by iwehrman** [comment]
This could be due to the different UI size on a big 5k iMac, but I find the HUD to be a little too small compared to the rest of the UI:
image
In particular, it seems like the icons should roughly the same size as the toolbar icons? CC @placegraphichere

**Comment by iwehrman** [comment]
If the HUD is up and I switch tools then switch back to the sampler, the HUD is still up with the wrong colors.

**Comment by iwehrman** [comment]
Currently sampling the stroke with the HUD sets the fill of the target layer. @placegraphichere thinks that it should set the stroke of the target layer. Fight.

**Comment by iwehrman** [comment]
Also, re: hole in the icon/icons are too small: the Fx icon is also hard to click. Maybe I'm just bad at clicking.

**Comment by iwehrman** [comment]
If I hold down the command key and move the mouse, the cursor turns into a move arrow. That ain't right.

File: src/js/actions/sampler.js [hunk 9]
```diff
@@ -176,7 +289,64 @@ define(function (require, exports) {
...
+                var hitLayerMap = new Set(hitLayerIDs.toJS()),
+                    clickedLeafLayers = layerTree.leaves.filter(function (layer) {
+                        return hitLayerMap.has(layer.id);
+                    }),
+                    topLayer = clickedLeafLayers.last();
+
+                return topLayer;
+            })
+            .then(function (source) {
+                return _calculateSampleTypes.call(this, doc, source, x, y);
```
Comment (iwehrman) [inline src/js/actions/sampler.js]: This can be moved into the previous then handler.

File: src/js/actions/sampler.js [hunk 10]
```diff
@@ -197,9 +367,37 @@ define(function (require, exports) {
...
             return Promise.resolve();
         }
+
+        var fillColor = null,
+            strokeColor = null,
+            typeStyle = null;
+
+        switch (source.kind) {
+        case source.layerKinds.VECTOR:
+            fillColor = source.fill ? source.fill.color : null;
```
Comment (iwehrman) [inline src/js/actions/sampler.js]: Could be just `fillColor = source.fill && source.fill.color`.

**Comment by iwehrman** [comment]
OK, done with this pass. Looks cool! Big questions are:

1. correct behavior of the stroke button?
2. keydown instead of keyup?
3. bigger/easier click targets?

File: src/js/models/layer.js (hunk 7 above)
Comment (volfied) [inline src/js/models/layer.js]: I thought about that... but these are values that we defined, unlike layer kinds.

**Comment by volfied** [comment]
@iwehrman Made the review changes. 

Re: stroke behavior: this is annoying to me as well.. it feels like shape->shape stroke sampling should sample... well, stroke! But it still applies to the main color of text layers, so it might be unexpected. I'd like to leave that one on the back burner, and see how it feels to QE and testers.

**Comment by iwehrman** [comment]
I'm not qualified to judge, but unless I'm mistaken @placegraphichere said he thought stroke should go to stroke, and that he and you discussed that. It's fine if it changes later, but I'd like to at least get everyone on the same page about what the intended design is.

**Comment by volfied** [comment]
I must have remembered wrong. It seems to make sense to everyone, so I'll make the change!

**Comment by iwehrman** [comment]
Filed #2135 about the change in behavior for the stroke button. Looks good otherwise, thanks!

**Pull request merged**